### PR TITLE
3.0 - Improved belongsToMany select field support

### PR DESCRIPTION
**Pull request by ndm2**
I'm not too sure about the magic options, as 3.x was supposed to have less magic, however it's just a tiny bit of magic, and since similar is supported for belongsTo associations, this might be within reason.

### CODE REVIEW

File: src/View/Helper/FormHelper.php [hunk 1]
```diff
@@ -1167,9 +1170,14 @@ protected function _optionsOptions($fieldName, $options)
             return $options;
         }
 
+        if (substr($fieldName, -5) === '._ids') {
+            $fieldName = substr($fieldName, 0, -5);
+        } elseif (substr($fieldName, -3) === '_id') {
+            $fieldName = substr($fieldName, 0, -3);
+        }
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: Why is this necessary? Line 1170 currently takes the first segment which would be `tags` in the case of `tags._ids`.

File: src/View/Helper/FormHelper.php [hunk 2]
```diff
@@ -1167,9 +1170,14 @@ protected function _optionsOptions($fieldName, $options)
...
         $fieldName = array_slice(explode('.', $fieldName), -1)[0];
         $varName = Inflector::variable(
-            Inflector::pluralize(preg_replace('/_id$/', '', $fieldName))
+            Inflector::pluralize($fieldName)
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: This isn't going to work for many scenarios as there are a huge pile of words that cannot be pluralized twice. I think we're just getting 'lucky' with tags.  I'd rather have no magic vs. inconsistent magic.

File: src/View/Helper/FormHelper.php (hunk 1 above)
Comment (ndm2) [inline src/View/Helper/FormHelper.php]: Actually it takes the last segment.

File: src/View/Helper/FormHelper.php (hunk 2 above)
Comment (ndm2) [inline src/View/Helper/FormHelper.php]: Ah, right, I vaguely remember that this caused problems here and there before :)

Comment (ndm2) [inline src/View/Helper/FormHelper.php]: Manually changed association properties and foreign keys aside - As per conventions, it should only be necessary to pluralize belongsTo/hasOne fields, shouldn't it? If that is the case, could pluralizing based on the postfix be a viable solution (given that it's worth to keep the magic anyways)?

Comment (markstory) [inline src/View/Helper/FormHelper.php]: Pluralizing based on `_id` but not `._ids` seems within reason to me.

Comment (ndm2) [inline src/View/Helper/FormHelper.php]: Meh, looking at the tests, namely [`testInputOverridingMagicSelectType`](https://github.com/ndm2/cakephp/blob/3.0-improved-belongs-to-many-select-field-support/tests/TestCase/View/Helper/FormHelperTest.php#L2934) and [`testInputMagicSelectForTypeNumber`](https://github.com/ndm2/cakephp/blob/3.0-improved-belongs-to-many-select-field-support/tests/TestCase/View/Helper/FormHelperTest.php#L3033), it seems that the general pluralization to magically pick up any variable based on this convention is still wanted?

Comment (markstory) [inline src/View/Helper/FormHelper.php]: Wanted-ish. I think I tried to remove it at some point and people were not happy about that.

Comment (ndm2) [inline src/View/Helper/FormHelper.php]: I see... guess I'll have to go with only excluding `._ids` postfixed fieldnames then.

File: src/View/Helper/FormHelper.php [hunk 3]
```diff
@@ -674,6 +674,9 @@ public function isFieldError($field)
      */
     public function error($field, $text = null, array $options = [])
     {
+        if (substr($field, -5) === '._ids') {
+            $field = substr($field, 0, -5);
+        }
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: Will the array property ever have a validation error attached to it?

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 4]
```diff
@@ -4126,6 +4126,79 @@ public function testHabtmSelectBox()
             '/div'
         ];
         $this->assertHtml($expected, $result);
+
+        $tags = [
+            1 => 'blue',
+            50 => 'green'
+        ];
+        $this->View->viewVars['tags'] = $tags;
```
Comment (markstory) [inline tests/TestCase/View/Helper/FormHelperTest.php]: It might be better to use something like `people` here as that is a name that would break if passed through pluralize().

File: src/View/Helper/FormHelper.php (hunk 3 above)
Comment (ndm2) [inline src/View/Helper/FormHelper.php]: It might have, say for example you wanted to validate that a maximum/minimum of x tags is being selected.

File: tests/TestCase/View/Helper/FormHelperTest.php (hunk 4 above)
Comment (ndm2) [inline tests/TestCase/View/Helper/FormHelperTest.php]: Indeed

**Comment by markstory** [comment]
Looks good to me.
